File: homeworks/fall2021/python/SortingDriver.py

from Sorting import insertionSort, quickSort
import copy
import random

def checkSorted(orig, sorted):

	# Make sure size is the same
	if len(orig) != len(sorted):
		print("ERROR...original list and sorted list have different lengths...",end='')
		return False
	
	# Make sure new array is sorted
	for i in range(1,len(sorted)):
		if sorted[i] < sorted[i-1]:
			print("ERROR...the sorted list does not appear to be correctly sorted...",end='')
			return False

	# Make sure the same elements are in each array
	count = {}
	for i in range(len(orig)):
		if orig[i] not in count:
			count[orig[i]] = 1
		else:
			count[orig[i]] = count[orig[i]]+1
		
		if sorted[i] not in count:
			count[sorted[i]] = -1
		else:
			count[sorted[i]] = count[sorted[i]]-1;

	for key in count:
		if count[key] != 0:
			print("ERROR...The sorted list does not contain the same elements that the original list does...",end='')
			return False
	
	return True



SIZE = 100
	
# Make an array (vector) to sort. Fill with random numbers
list = []
for i in range(SIZE):
	list.append((int)(random.randrange(SIZE)))

# Make copies to sort
ins = copy.deepcopy(list) 
qui = copy.deepcopy(list)

print("Sorting using insertion sort...",end='')
insertionSort(ins, 0, len(list)-1);
print("DONE\nChecking if sorted correctly...",end='')
checkSorted(list, ins);
print("DONE")

print("Sorting using quick sort...",end='')
quickSort(qui, 0, len(list)-1, 1);
print("DONE\nChecking if sorted correctly...",end='')
checkSorted(list, qui);
print("DONE")

File: homeworks/fall2021/cpp/SortingDriver.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cstdlib>

using namespace std;

void insertionSort(vector<int> &list, int start, int end);
void quickSort(vector<int> &list, int i, int j, int minSize);
bool checkSorted(vector<int> &orig, vector<int> &sorted);

/* Number of elements to sort */
int SIZE = 100;

int main() {	
	
	/* Make an array (vector) to sort. Fill with random numbers */
	vector<int> list;
	for(int i=0; i<SIZE; i++) list.push_back((int)(rand()*SIZE));
	
	/* Make copies to sort */
	vector<int> ins;
	vector<int> qui;

	ins = list; qui = list; //should make deep copies
	

	cout << "Sorting using insertion sort...";
	insertionSort(ins, 0, ins.size()-1);
	cout << "DONE\nChecking if sorted correctly...";
	checkSorted(list, ins);
	cout << "DONE" << endl;
	
	cout << "Sorting using quick sort...";
	quickSort(qui, 0, qui.size()-1, 1);
	cout << "DONE\nChecking if sorted correctly...";
	checkSorted(list, qui);
	cout << "DONE" << endl;

}

	
bool checkSorted(vector<int> &orig, vector<int> &sorted) {

	/* Make sure size is the same */
	if(orig.size() != sorted.size()) {
		cout << "ERROR...original list and sorted list have different lengths...";
		return false;
	}
	
	/* Make sure new array is sorted */
	for(int i=1; i<sorted.size(); i++) {
		if(sorted[i] < sorted[i-1]) {
			cout << "ERROR...the sorted list does not appear to be correctly sorted...";
			return false;
		}
	}
	
	/* Make sure the same elements are in each array */
	unordered_map<int, int> count;
	for(int i=0; i<orig.size(); i++) {
		if(count.count(orig[i]) == 0) count[orig[i]] = 1;
		else count[orig[i]]++;
		
		if(count.count(sorted[i]) == 0) count[sorted[i]] = -1;
		else count[sorted[i]]--;
	}
	for(auto &key : count) {
		if(key.second != 0) {
			cout << "ERROR...The sorted list does not contain the same elements that the original list does...";
			return false;
		}
	}
	
	return true;
}

/**
 * Use this printlist method while debugging if helpful
 * @param list
 */
void printList(vector<int> &list) {
	for(int i=0; i<list.size(); i++) {
		cout << list[i] << ", ";
	}
	cout << endl;
}
